File: scripts/check-agents-links.py
#!/usr/bin/env python3
"""Validate local markdown links in AGENTS documentation only.

Checks:
- root AGENTS.md
- libs/*/AGENTS.md

Exits non-zero when broken local links are found.
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
LINK_RE = re.compile(r"\[[^\]]+\]\(([^)]+)\)")


def iter_agents_files(root: Path) -> list[Path]:
    files = [root / "AGENTS.md"]
    files.extend(sorted((root / "libs").glob("*/AGENTS.md")))
    return [f for f in files if f.exists()]


def is_skippable(link: str) -> bool:
    return link.startswith(("http://", "https://", "mailto:", "#"))


def resolve_target(base_file: Path, link: str) -> Path:
    target = link.split("#", 1)[0]
    if target.startswith("/"):
        return Path(target)
    return (base_file.parent / target).resolve()


def main() -> int:
    files = iter_agents_files(ROOT)
    broken: list[tuple[Path, str, Path]] = []
    checked_links = 0

    for file in files:
        content = file.read_text(encoding="utf-8")
        for match in LINK_RE.finditer(content):
            link = match.group(1).strip()
            if is_skippable(link):
                continue
            checked_links += 1
            target = resolve_target(file, link)
            if not target.exists():
                broken.append((file, link, target))

    print(f"AGENTS files checked: {len(files)}")
    print(f"Local links checked: {checked_links}")
    print(f"Broken links: {len(broken)}")

    if broken:
        for file, link, target in broken:
            rel = file.relative_to(ROOT)
            print(f"\n{rel}\n  -> {link}\n  => {target}")
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/check_agents_links.clj
#!/usr/bin/env bb
;; scripts/check_agents_links.clj
;;
;; Validate local markdown links in AGENTS documentation only.
;;
;; Checks:
;; - root AGENTS.md
;; - libs/*/AGENTS.md
;;
;; Exits non-zero when broken local links are found.
;;
;; Usage (babashka):
;;   bb scripts/check_agents_links.clj
;;   bb check-links              (via bb.edn task)

(ns check-agents-links
  (:require [clojure.string :as str]
            [clojure.java.io :as io]))

(def root-dir (io/file (System/getProperty "user.dir")))

(defn iter-agents-files []
  (let [root-agents (io/file root-dir "AGENTS.md")
        libs-dir (io/file root-dir "libs")
        lib-agents (when (.exists libs-dir)
                     (->> (.listFiles libs-dir)
                          (filter #(.isDirectory %))
                          (sort-by #(.getName %))
                          (map #(io/file % "AGENTS.md"))))]
    (->> (cons root-agents lib-agents)
         (filter #(.exists %)))))

(defn skippable? [link]
  (or (str/starts-with? link "http://")
      (str/starts-with? link "https://")
      (str/starts-with? link "mailto:")
      (str/starts-with? link "#")))

(defn resolve-target [base-file link]
  (let [target (first (str/split link #"#"))]
    (if (str/starts-with? target "/")
      (io/file root-dir (subs target 1))
      (.getCanonicalFile (io/file (.getParentFile base-file) target)))))

(defn check-file [file]
  (let [content (slurp file)
        link-pattern #"\[[^\]]+\]\(([^)]+)\)"
        local-links (->> (re-seq link-pattern content)
                         (map second)
                         (map str/trim)
                         (remove skippable?))
        broken (->> local-links
                    (map (fn [link]
                           (let [target (resolve-target file link)]
                             (when-not (.exists target)
                               {:file file :link link :target target}))))
                    (remove nil?))]
    {:checked (count local-links)
     :broken broken}))

(defn -main []
  (let [files (vec (iter-agents-files))
        results (map check-file files)
        total-checked (reduce + (map :checked results))
        all-broken (vec (mapcat :broken results))]
    (println (str "AGENTS files checked: " (count files)))
    (println (str "Local links checked: " total-checked))
    (println (str "Broken links: " (count all-broken)))
    (when (seq all-broken)
      (doseq [{:keys [file link target]} all-broken]
        (let [rel (.relativize (.toPath root-dir) (.toPath file))]
          (println (str "\n" rel "\n  -> " link "\n  => " (.getPath target))))))
    (if (seq all-broken) 1 0)))

;; Run when executed directly
(when (= *file* (System/getProperty "babashka.file"))
  (System/exit (-main)))
